rig/machine.py

```diff
@@ -0,0 +1,220 @@
+    described below. A number of utility methods are available but should be
+    considered just that: utilities.
+
+    Attributes
+    ----------
+    width : int
+        The width of the system in chips: chips will thus have x-coordinates
+        between 0 and width-1 inclusive.
+    height : int
+        The height of the system in chips: chips will thus have y-coordinates
+        between 0 and height-1 inclusive.
+    chip_resources : {resource_key: requirement, ...}
+        The resources available on chips (unless otherwise stated in
+        `chip_resource_exceptions). `resource_key` must be some unique
+        identifying object for a given resource. `requirement` must be a
+        positive numerical value.
+    chip_resource_exceptions : {(x,y): resources, ...}
+        If any chip's resources differ from those specified in
+        `chip_resources`, an entry in this dictionary with the key being the
+        chip's coordinates as a tuple `(x, y)` and `resources` being a
+        dictionary of the same format as `chip_resources`. Note that every
+        exception must specify exactly the same set of keys as
+        `chip_resources`.
+    dead_chips : set
+        A set of `(x,y)` tuples enumerating all chips which completely
+        unavailable. Links leaving a dead chip are implicitly marked as dead.
+    dead_links : set
+        A set `(x,y,link)` where `x` and `y` are a chip's coordinates and
+        `link` is a value from the Enum :py:class:`~rig.machine.Links`. Note
+        that links have two directions and both should be defined if a link is
+        dead in both directions (the typical case).
+    """
+    __slots__ = ["width", "height", "chip_resources",
+                 "chip_resource_exceptions", "dead_chips", "dead_links"]
+
+    def __init__(self, width, height,
+                 chip_resources={Cores: 18, SDRAM: 128*1024*1024,
+                                 SRAM: 32*1024},
+                 chip_resource_exceptions={}, dead_chips=set(),
+                 dead_links=set()):
+        """Defines the resources available within a SpiNNaker system.
+
+        Parameters
+        ----------
+        width : int
+        height : int
+        chip_resources : {resource_key: requirement, ...}
+        chip_resource_exceptions : {(x,y): resources, ...}
+        dead_chips : set([(x,y,p), ...])
+        dead_links : set([(x,y,link), ...])
+        """
+        self.width = width
+        self.height = height
+
+        self.chip_resources = chip_resources.copy()
+        self.chip_resource_exceptions = chip_resource_exceptions.copy()
+
+        self.dead_chips = dead_chips.copy()
+        self.dead_links = dead_links.copy()
+
+    def copy(self):
+        """Produce a copy of this datastructure."""
+        return Machine(
+            self.width, self.height,
+            self.chip_resources, self.chip_resource_exceptions,
+            self.dead_chips, self.dead_links)
```

Comment:
The copying is done in the constructor (/me is copying the convention you may have started in the Net container :)).

The use case is that some constraints (e.g. reserved resource constraints) are easily modelled to a placer by simply reducing the total resources available by the reserved amount. The most straight-forward thing to do would be to copy the Machine and modify it accordingly rather than having to check if a constraint exists and if not, looking at the machine to determine resource availability.